internal/langserver/handlers/command/module_providers.go

```diff
@@ -0,0 +1,69 @@
+package command
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/creachadair/jrpc2/code"
+	lsctx "github.com/hashicorp/terraform-ls/internal/context"
+	"github.com/hashicorp/terraform-ls/internal/langserver/cmd"
+	"github.com/hashicorp/terraform-ls/internal/uri"
+)
+
+const moduleProvidersVersion = 0
+
+type moduleProvidersResponse struct {
+	FormatVersion        int                            `json:"v"`
+	ProviderRequirements map[string]providerRequirement `json:"provider_requirements"`
+	InstalledProviders   map[string]string              `json:"installed_providers"`
+}
+
+type providerRequirement struct {
+	DisplayName       string `json:"display_name"`
+	VersionConstraint string `json:"version_constraint"`
+}
+
+func ModuleProvidersHandler(ctx context.Context, args cmd.CommandArgs) (interface{}, error) {
+	response := moduleProvidersResponse{
+		FormatVersion:        moduleProvidersVersion,
+		ProviderRequirements: make(map[string]providerRequirement),
+		InstalledProviders:   make(map[string]string),
+	}
+
+	modUri, ok := args.GetString("uri")
+	if !ok || modUri == "" {
+		return response, fmt.Errorf("%w: expected module uri argument to be set", code.InvalidParams.Err())
+	}
+
+	if !uri.IsURIValid(modUri) {
+		return response, fmt.Errorf("URI %q is not valid", modUri)
+	}
+
+	modPath, err := uri.PathFromURI(modUri)
+	if err != nil {
+		return response, err
+	}
+
+	mm, err := lsctx.ModuleFinder(ctx)
+	if err != nil {
+		return response, err
+	}
+
+	found, _ := mm.ModuleByPath(modPath)
+	if found == nil {
+		return response, nil
+	}
+
+	for provider, version := range found.Meta.ProviderRequirements {
```

Comment:
I think we'll need to account for the possibility of the module not being parsed (yet) here, as this request may be called before the background operation collecting the data finishes. I guess that a simple `nil` check on `.Meta` may do the job for now.

In the long run we'll also address the lack of data by triggering update of the data from server to client. I will create a separate issue for that as we need to address the same problem in `module.calls`.

Reply:
Meta is a `ModuleMetadata` struct. Can a struct be `nil`? How check for that?

Reply:
Oops, sorry for the confusion - I missed that fact entirely, I thought it was a pointer.

In that case we could either turn it into a pointer, or - easier/quicker solution in the context of this PR - we can check that `MetaState != op.OpStateUnknown`. The parsing may still fail (in which case `MetaState = OpStateLoaded`), so if we would also need to check for `MetaErr == nil`.

The only downside with introducing a pointer is that we'd need to carefully check all references and make sure these also contain nil checks, so that we don't introduce a crash.

Reply:
Okay, I've added a check for `MetaState`  and `MetaErr`. This seems a lot easier for now than converting `ModuleMetadata` to a pointer.